contracts/TrustedForwarder.sol

```diff
@@ -0,0 +1,40 @@
+pragma solidity ^0.5.16;
+pragma experimental ABIEncoderV2;
+
+import "./utils/GSNTypes.sol";
+import "./utils/EIP712Sig.sol";
+import "./interfaces/ITrustedForwarder.sol";
+
+contract TrustedForwarder is ITrustedForwarder {
+
+    EIP712Sig private eip712sig;
+
+    // Nonces of senders, used to prevent replay attacks
+    mapping(address => uint256) private nonces;
+
+    constructor() public {
+        eip712sig = new EIP712Sig(address(this));
+    }
+
+    function getNonce(address from) external view returns (uint256) {
+        return nonces[from];
+    }
+
+    function verify(GSNTypes.RelayRequest memory req, bytes memory sig) view public returns (bool){
```

Comment:
I do not like that there is a bi-directional API between Forwarder and Relay Hub. This means there is a need for all participants to trust each other and communicate.
A malicious forwarder can drain a Paymaster by always returning `true` but reverting all transactions.
Relay Hub should not express any guarantees about forwarder. With this structure, the Paymaster will need to decide if it wants to trust this Forwarder's `verify` method.
This may be acceptable, but can we design around it?